libbpf-rs/src/tc.rs

```diff
@@ -0,0 +1,291 @@
+    /// if unset upon attach, the kernel will assign a priority for the hook
+    pub fn priority(&mut self, priority: u32) -> &mut Self {
+        self.opts.priority = priority;
+        self
+    }
+
+    /// Query a hook to inspect the program identifier (prog_id)
+    pub fn query(&mut self) -> Result<u32> {
+        let mut opts = self.opts;
+        opts.prog_id = 0;
+        opts.prog_fd = 0;
+        opts.flags = 0;
+
+        let err = unsafe { libbpf_sys::bpf_tc_query(&self.hook as *const _, &mut opts as *mut _) };
+        if err != 0 {
+            Err(Error::System(errno::errno()))
+        } else {
+            Ok(opts.prog_id)
+        }
+    }
+
+    /// Attach the TcHook so that the program starts processing
+    ///
+    /// Once the hook is processing, changing the values will have no effect
+    /// unless the hook is attach()'d again (replace(true) being required)
+    ///
+    /// Users can create a second hook by changing the handle, the priority
+    /// or the attach_point and calling the attach() method again.
+    /// Beware doing this.  It might be better to Copy the TcHook and change
+    /// the values on the copied hook for easier detach()
+    ///
+    /// NOTE: Once a TcHook is attached, it, and the maps it uses,
+    ///       will outlive the userspace application that spawned them
+    ///       Make sure to detach if this is not desired
+    pub fn attach(&mut self) -> Result<Self> {
+        self.opts.prog_id = 0;
+        let err =
+            unsafe { libbpf_sys::bpf_tc_attach(&self.hook as *const _, &mut self.opts as *mut _) };
+        if err != 0 {
+            Err(Error::System(errno::errno()))
+        } else {
+            Ok(*self)
+        }
+    }
+
+    /// Detach a TcHook
+    pub fn detach(&mut self) -> Result<()> {
+        let mut opts = self.opts;
+        opts.prog_id = 0;
+        opts.prog_fd = 0;
+        opts.flags = 0;
+
+        let err = unsafe { libbpf_sys::bpf_tc_detach(&self.hook as *const _, &opts as *const _) };
+        if err != 0 && err != -(errno::Errno::ENOENT as i32) {
+            Err(Error::System(err as i32))
+        } else {
+            self.opts.prog_id = 0;
+            Ok(())
+        }
+    }
+
+    /// Destroy attached filters
+    ///
+    /// If called on a hook with an attach_point of EGRESS, will detach all EGRESS hooks
+    /// If called on a hook with an attach_point of INGRESS, will detach all INGRESS hooks
+    ///
+    /// If called on a hook with an attach_point of EGRESS|INGRESS, will destroy the clsact tc
+    /// qdisc and detach all hooks
+    ///
+    /// Will error with EOPNOTSUPP if attach_point is BPF_TC_CUSTOM
+    ///
+    /// It is good practice to query before destroying as the tc qdisc may be used
+    /// by multiple programs
+    pub fn destroy(&mut self) -> Result<()> {
+        let err = unsafe { libbpf_sys::bpf_tc_hook_destroy(&mut self.hook as *mut _) };
+        if (err == -(EINVAL as i32) || err == -(ENOENT as i32))
```

Comment:
Same here - I'm not seeing where the path to ENOENT is in netlink.c for bpf_tc_hook_destroy. Where do you see this?

Reply:
https://elixir.bootlin.com/linux/latest/source/net/sched/sch_api.c#L1485

Resolved by returning the errors back to the user.